pkg/util/limiter/memory_limiter.go

```diff
@@ -0,0 +1,242 @@
+package limiter
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"go.uber.org/atomic"
+)
+
+var (
+	errCheckIntervalOutOfRange = errors.New(
+		"checkInterval must be greater than zero")
+	errLimitOutOfRange = errors.New(
+		"MemoryLimitBytes or ResetBytes must be greater than zero")
+)
+
+type MemoryLimiterConfig struct {
+	Enabled          bool          `yaml:"enabled"`
+	MemoryLimitBytes uint64        `yaml:"memory_limit_bytes"`
+	ResetBytes       uint64        `yaml:"reset_bytes"`
+	CheckInterval    time.Duration `yaml:"check_interval"`
+}
+
+// RegisterFlags registers flags.
+func (cfg *MemoryLimiterConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.BoolVar(&cfg.Enabled, prefix+"memory-limiter.enabled", false, "Enable memory-limiter.")
+	f.DurationVar(&cfg.CheckInterval, prefix+"memory-limiter.check-interval", time.Second, "Configures how often to watch the heap to detect memory pressure.")
+	f.Uint64Var(&cfg.MemoryLimitBytes, prefix+"memory-limiter.memory-limit-bytes", 0, "Specifies the memory limit above which memory is considered to be under pressure.")
+	f.Uint64Var(&cfg.ResetBytes, prefix+"memory-limiter.reset-bytes", 0, "Specifies the memory limit below which full operation is resumed.")
+}
+
+// Interface for the memory limiter.
```

Comment:
This comment is redundant; I think the a comment about how to use this interface would be more useful.

Reply:
Updated.